File: leetcode/dp/3610_Minimum_Number_of_Primes_to_Sum_to_Target/codes/eratosthenes.c
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

int minNumberOfPrimes(int n, int m) {
    int primes[1000], count = 0;
    int size = 8000;
    bool isPrime[8000];
    for (int i = 0; i < size; ++i) isPrime[i] = true;
    isPrime[0] = isPrime[1] = false;

    for (int i = 2; i * i < size; ++i) {
        if (isPrime[i]) {
            for (int j = i * i; j < size; j += i)
                isPrime[j] = false;
        }
    }

    for (int i = 2; i < size && count < m; ++i) {
        if (isPrime[i])
            primes[count++] = i;
    }

    int dp[1001];
    for (int i = 0; i <= n; ++i)
        dp[i] = INT_MAX;
    dp[0] = 0;

    for (int i = 0; i < m; ++i) {
        for (int j = primes[i]; j <= n; ++j) {
            if (dp[j - primes[i]] != INT_MAX && dp[j - primes[i]] + 1 < dp[j])
                dp[j] = dp[j - primes[i]] + 1;
        }
    }

    return dp[n] == INT_MAX ? -1 : dp[n];
}

File: leetcode/dp/3610_Minimum_Number_of_Primes_to_Sum_to_Target/codes/eratosthenes.py
from typing import List

class Solution:
    def minNumberOfPrimes(self, n: int, m: int) -> int:
        def generate_primes(limit_count: int) -> List[int]:
            size = 8000  
            is_prime = [True] * size
            is_prime[0] = is_prime[1] = False

            for i in range(2, int(size ** 0.5) + 1):
                if is_prime[i]:
                    for j in range(i * i, size, i):
                        is_prime[j] = False

            primes = [i for i, val in enumerate(is_prime) if val]
            return primes[:limit_count]

        primes = generate_primes(m)
        dp = [float("inf")] * (n + 1)
        dp[0] = 0

        for p in primes:
            for i in range(p, n + 1):
                if dp[i - p] + 1 < dp[i]:
                    dp[i] = dp[i - p] + 1

        return dp[n] if dp[n] != float("inf") else -1
